clang/lib/Driver/ToolChains/Repo.cpp

```diff
@@ -0,0 +1,198 @@
+//===--- Repo.cpp - Repo ToolChain Implementations --------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Repo.h"
+#include "CommonArgs.h"
+#include "clang/Driver/Compilation.h"
+#include "clang/Driver/DriverDiagnostic.h"
+#include "llvm/Support/VirtualFileSystem.h"
+
+using namespace clang::driver;
+using namespace clang::driver::tools;
+using namespace clang::driver::toolchains;
+using namespace clang;
+using namespace llvm::opt;
+
+static void constructRepoMuslLinkArgs(Compilation &C, const JobAction &JA,
+                                      const toolchains::RepoMuslToolChain &RTC,
+                                      const InputInfo &Output,
+                                      const InputInfoList &Inputs,
+                                      const ArgList &Args,
+                                      ArgStringList &CmdArgs,
+                                      const char *LinkingOutput) {
+
+  const Driver &D = RTC.getDriver();
+
+  //----------------------------------------------------------------------------
+  //
+  //----------------------------------------------------------------------------
+
+  const std::string MuslRoot = D.SysRoot.empty() ? "/usr/local/musl" : D.SysRoot;
+  if (!Args.hasArg(options::OPT_nostartfiles, options::OPT_nostdlib)) {
+    CmdArgs.push_back(Args.MakeArgString(MuslRoot + "/lib/crt1.t"));
+    CmdArgs.push_back(Args.MakeArgString(MuslRoot + "/lib/crt1_asm.t"));
+    CmdArgs.push_back(Args.MakeArgString(MuslRoot + "/lib/libc_repo.a"));
+  }
+
+  // If compiler-rt is built inside of LLVM project, we could find its path
+  // using getCompilerRTPath function. If build compiler-rt as stand-alone
+  // project and give --sysroot=preference, use sysroot as the libary root.
+  auto CRTPath = RTC.getCompilerRTPath();
+  if (!RTC.getVFS().exists(CRTPath))
+    CRTPath = D.SysRoot.empty() ? "/usrlib/linux" : D.SysRoot + "/lib/linux/";
+  if (!Args.hasArg(options::OPT_nostartfiles, options::OPT_nostdlib)) {
+    CmdArgs.push_back(
+        Args.MakeArgString(CRTPath + "/clang_rt.crtbegin-x86_64.o"));
```

Comment:
Elsewhere, the code uses `llvm::sys::path::append()` to join path strings, but here (as well as in a few other places) we just concatenate the strings. This seems rather inconsistent.

Reply:
I have changed to use the `llvm::sys::path::append()`. However, for the string passing to `Args.MakeArgString` function, I have to add two variables `CRTBegin` and `CRTEnd`. These two variables are only used one time. For the arguments of `Args.MakeArgString`, should we still used concatenate the strings (like other LLVM files/codes)? 

>
> as well as in a few other places.
>
Other places are mainly for the arguments of `Args.MakeArgString`, I still used concatenate the strings (like other LLVM files/codes).